tests/mock_crust/mod.rs

```diff
@@ -213,77 +208,144 @@ fn multiple_clients_per_proxy() {
         }
-        trace!("clients_sent: {:?}", clients_sent);
-        let _ = poll_all(&mut nodes, &mut clients);
-
-        let mut request_received: HashMap<MessageId, usize> = HashMap::new();
-        for node in nodes.iter_mut().filter(|n| n.is_recipient(&dst)) {
-            while let Ok(event) = node.try_next_ev() {
-                if let Event::Request {
-                           request: Request::GetIData { msg_id: req_message_id, .. }, ..
-                       } = event {
-                    let entry = request_received
-                        .entry(req_message_id)
-                        .or_insert_with(|| 0);
-                    *entry += 1;
-                }
+    }
+    trace!("clients_sent: {:?}", clients_sent);
+    let _ = poll_all(nodes, clients);
+
+    let mut request_received: HashMap<MessageId, usize> = HashMap::new();
+    for node in nodes.iter_mut().filter(|n| n.is_recipient(&dst)) {
+        while let Ok(event) = node.try_next_ev() {
+            if let Event::Request {
+                       request: Request::GetIData { msg_id: req_message_id, .. }, ..
+                   } = event {
+                let entry = request_received
+                    .entry(req_message_id)
+                    .or_insert_with(|| 0);
+                *entry += 1;
             }
         }
-        trace!("request_received: {:?}", request_received);
+    }
+    trace!("request_received: {:?}", request_received);
+
+    for (msg_id, count) in &request_received {
+        assert_eq!(*count, MIN_SECTION_SIZE);
+        let _ = unwrap!(clients_sent.remove(msg_id));
+        *total_usage += MAX_IMMUTABLE_DATA_SIZE_IN_BYTES;
+    }
+    assert!(*total_usage <= CAPACITY);
+
+    clients_sent
+}
 
-        for (msg_id, count) in &request_received {
-            assert_eq!(*count, min_section_size);
-            let _ = unwrap!(clients_sent.remove(msg_id));
-            total_usage += MAX_IMMUTABLE_DATA_SIZE_IN_BYTES;
+// Verifies the usage in rate_limiter. Also advances the fake clock.
+fn rate_limiter_verify(clients_sent: &HashMap<MessageId, IpAddr>,
+                       clients_usage: &BTreeMap<IpAddr, u64>,
+                       total_usage: &mut u64,
+                       per_client_cap: u64) {
+    // `clients_sent` now contains only the clients whose request got rejected.
+    // Needs to confirm such rejection is valid. However, if it is the total usage reaching the
+    // cap, the usage of each client could be much less than the cap.
+    if (*total_usage + MAX_IMMUTABLE_DATA_SIZE_IN_BYTES) <= CAPACITY {
+        for ip in clients_sent.values() {
+            assert!((unwrap!(clients_usage.get(ip)) + MAX_IMMUTABLE_DATA_SIZE_IN_BYTES) >
+                        per_client_cap);
         }
-        assert!(total_usage <= CAPACITY);
+    }
+
+    let wait_millis = 2 * MAX_IMMUTABLE_DATA_SIZE_IN_BYTES * 1000 / RATE as u64;
+    let leaky_rate = 2 * MAX_IMMUTABLE_DATA_SIZE_IN_BYTES;
```

Comment:
I think it might be clearer and more robust to define the desired leak amount first, then derive the wait duration from it; e.g.

```rust
let leaked_units = 2 * MAX_IMMUTABLE_DATA_SIZE_IN_BYTES;
let wait_millis = (leaked_units * 1000) / RATE as u64;
```